python/composio/tools/toolset.py

```diff
@@ -779,6 +784,18 @@ def execute_action(
         :return: Output object from the function call
         """
         action = Action(action)
+        if (
+            self.requested_actions is not None
+            and self.requested_apps is not None
+            and action.slug not in self.requested_actions
+            and action.app not in self.requested_apps
+        ):
+            raise ComposioSDKError(
+                f"Action {action.slug} is being called, but was never requested by the toolset."
```

Comment:
The error message could be more informative. Consider including:
1. The list of allowed actions/apps that were requested
2. A clearer separation between the error description and the explanation
3. A reference to documentation

Suggested format:
```python
raise ComposioSDKError(
    f"Action '{action.slug}' cannot be executed - not in requested tools.\n"
    f"Allowed actions: {self.requested_actions}\n"
    f"Allowed apps: {self.requested_apps}\n"
    "See: https://docs.composio.dev/validation for more details"
)
```